### PR TITLE
fix some audit issues

### PR DESCRIPTION
**Pull request by lispc**
TODO(future pr):

1. snark verifier / ecc / halo2-libs issues
2. call gas cost: revisit after precompiles
3. rlp u64 lt_128
4. SameContextGadget for some creation errors: ?revisit after  creation merged in upstream?

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs [hunk 1]
```diff
@@ -31,6 +31,7 @@ impl<F: Field, const N_BYTES: usize> ConstantDivisionGadget<F, N_BYTES> {
         numerator: Expression<F>,
         denominator: u64,
     ) -> Self {
+        assert!(N_BYTES <= 31);
```
Comment (naure) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs]: This is not the same as `lt.rs`. A complete overflow check would be this:

```rust
    assert!(N_BYTES * 8 + 64 - denominator.leading_zeros() as usize <= MAX_N_BYTES_INTEGER * 8);
```

File: zkevm-circuits/src/witness/rw.rs [hunk 2]
```diff
@@ -274,8 +274,8 @@ impl<F: Field> RwRow<F> {
     }
     pub(crate) fn rlc(&self, randomness: F) -> F {
         let values = self.values();
-        values
-            .iter()
+        std::iter::once(&F::one())
```
Comment (naure) [inline zkevm-circuits/src/witness/rw.rs:277]: I don’t understand this change. To which issue is it related?

**Comment by naure** [comment]
# Summary of the changes

## [High] TOB-SCROLL-1 The ModGadget is underconstrained and allows proving incorrect MULMOD operations

Fix OK. In ModGadget, `n=0` implies `a_or_zero=0`, which combined imply `r=0.`. In MulModGadget, `n=0` implies `a_reduced=0` which implies `d=e=0` which implies `r=0`.

- Files: `modulo.rs` (see also `mulmod.rs`)
- Status: Fixed


## [High] TOB-SCROLL-3 The BLOCKHASH opcode is underconstrained and allows computing the hash of any block

The call to `block_lookup` is OK.

- Files: `blockhash.rs`
- Status: Fixed


## [Info] TOB-SCROLL-5 N_BYTES parameters are not checked to prevent overflow

- Files: `lt.rs`, `constant_division.rs`, and `range_check.rs`
- Status: Fixed


## [High] TOB-SCROLL-7 Underconstrained warm status on CALL opcodes allow gas cost forgery

Couldn’t we replace the cell `is_warm` with a constant `1`?

The initial value is currently not enforced. Waiting on some task because of precompile behavior ([FIXME](https://github.com/scroll-tech/zkevm-circuits/blob/99cb6a7b261b3761b8bed997c9c4543560ffe2c4/zkevm-circuits/src/state_circuit/constraint_builder.rs#L157)).

- Files: `callop.rs`
- Status: On Hold.


## [High] TOB-SCROLL-9 The CREATE and CREATE2 opcodes can be called within a static context

The constraint "is_static is false" was added. The witness generation accounts for the extra RW op, and it was simplified using `rw_offset`.

- Files: `zkevm-circuits/src/evm_circuit/execution/create.rs` and `bus-mapping/src/evm/opcodes/create.rs`
- Status: Fixed.


## [Info] TOB-SCROLL-10 ResponsibleOpcode table incorrectly handles CREATE and CREATE2

- Files: `step.rs`
- Status: Fixed.


## [Info] TOB-SCROLL-B Code Quality

### Use Transition::Same instead of Delta(0.expr())

- File: `not.rs`
- Status: Fixed.

**Review by naure — APPROVED** [review]
This is great, let’s merge it and continue in new PRs.

File: zkevm-circuits/src/witness/rw.rs (hunk 2 above)
Comment (lispc) [reply zkevm-circuits/src/witness/rw.rs:277]: oh there is a sanity check feature when CHECK_RW_LOOKUP enabled. the sanity check was broken, during debugging in this PR i fixed the sanity check here.

**Comment by lispc** [comment]
TOB-SCROLL-3 The BLOCKHASH opcode is underconstrained and allows computing the hash of any block: kunxian comfirmed that block hash already constrained inside public input circuit?
TOB-SCROLL-7 Underconstrained warm status on CALL opcodes allow gas cost forgery:  cb.require_true constraint is_warm to be 1.  You mean we did not constrain is_warm_prev? Yes.. it is a TODO later.

**Comment by naure** [comment]
Got it. Let’s merge this. I am tracking the TODO for TOB-SCROLL-7.

**Pull request merged**